File: src/types/personalization.ts
export type PersonalizationMode = "basic" | "full";

export type ReflectionCadence = "none" | "daily" | "weekly";

export type GoalKey =
  | "health"
  | "relationships"
  | "career"
  | "execution"
  | "performance"
  | "creativity"
  | "mindfulness"
  | "learning"
  | string; // Allow custom goals as strings

export interface LearningStyle {
  visual: number;
  auditory: number;
  kinesthetic: number;
}

export type LanguageIntensity = "soft" | "neutral" | "direct";

export interface DriftRule {
  enabled: boolean;
  after: string | null;
}

export interface UserSettings {
  user_id?: string;
  personalization_mode: PersonalizationMode;
  local_cache_enabled: boolean;
  cadence: ReflectionCadence;
  goals: GoalKey[];
  extra_goal?: string | null;
  // custom_goals: string[];
  learning_style: LearningStyle;
  session_length_minutes: 10 | 25 | 45;
  spiritual_prompts: boolean;
  bluntness: number;
  language_intensity: LanguageIntensity;
  logging_format: "freeform" | "structured" | "mixed";
  drift_rule: DriftRule;
  crisis_card?: string | null;
  persona_tag: PersonaTag;
  checkin_notifications?: boolean;
  missed_day_notifications?: boolean;
  updated_at?: string;
  created_at?: string;
}

export type PersonaTag =
  | "Architect"
  | "Explorer"
  | "Anchor"
  | "Accelerator"
  | "Generalist";

export interface PersonalizationBundle {
  profile: ProfileSnapshot;
  settings: UserSettings | null;
}

export interface ProfileSnapshot {
  id: string;
  timezone: string;
  onboarding_completed: boolean;
  updated_at: string | null;
  missed_day_count?: number;
  current_streak?: number;
  last_message_at?: string | null;
}

export interface PersonaSignalPayload {
  source: "onboarding" | "settings_update";
  rationale: string;
  changes: Record<string, any>;
}

export interface PersonalizationState
  extends Omit<UserSettings, "persona_tag"> {
  persona_tag?: PersonaTag;
}

export interface CachedPersonalization {
  settings: UserSettings;
  updated_at: string;
}

File: services/python-ai/app/schemas/personalization.py
from typing import Dict, List, Optional
from pydantic import BaseModel


class LearningStyle(BaseModel):
    visual: float
    auditory: float
    kinesthetic: float


class DriftRule(BaseModel):
    enabled: bool
    after: Optional[str] = None


class UserSettings(BaseModel):
    user_id: Optional[str] = None
    personalization_mode: str
    local_cache_enabled: bool
    cadence: str
    goals: List[str]
    extra_goal: Optional[str] = None
    learning_style: LearningStyle
    session_length_minutes: int
    spiritual_prompts: bool
    bluntness: int
    language_intensity: str
    logging_format: str
    drift_rule: DriftRule
    crisis_card: Optional[str] = None
    persona_tag: str
    checkin_notifications: Optional[bool] = None
    missed_day_notifications: Optional[bool] = None
    updated_at: Optional[str] = None
    created_at: Optional[str] = None


class PersonaSignalPayload(BaseModel):
    source: str
    rationale: str
    changes: Dict[str, object]


class ProfileSnapshot(BaseModel):
    id: str
    timezone: str
    onboarding_completed: bool
    updated_at: Optional[str] = None
    missed_day_count: Optional[int] = None
    current_streak: Optional[int] = None
    last_message_at: Optional[str] = None


class PersonalizationRuntime(BaseModel):
    user_settings: Optional[UserSettings] = None
    persona: Optional[str] = None
    cadence: str
    tone: str
    spiritual_on: bool
    bluntness: int
    privacy_gates: Dict[str, bool]
    crisis_rules: Dict[str, object]
    resolved_at: str


class PersonalizationBundle(PersonalizationRuntime):
    profile: ProfileSnapshot
    settings: Optional[UserSettings] = None


class PersonalizationState(BaseModel):
    personalization_mode: str
    local_cache_enabled: bool
    cadence: str
    goals: List[str]
    extra_goal: Optional[str] = None
    learning_style: LearningStyle
    session_length_minutes: int
    spiritual_prompts: bool
    bluntness: int
    language_intensity: str
    logging_format: str
    drift_rule: DriftRule
    crisis_card: Optional[str] = None
    persona_tag: Optional[str] = None
    checkin_notifications: Optional[bool] = None
    missed_day_notifications: Optional[bool] = None
    updated_at: Optional[str] = None
    created_at: Optional[str] = None


class CachedPersonalization(BaseModel):
    settings: UserSettings
    updated_at: str
